Enable test_ellipsis_index_2 with Torch dynamo

laithsakka opened the pull request:
Fix issue #118819

test_ellipsis_index_2 is specifically testing properties of torch._numpy.array()
and that a field tensor is being added hence overriding the imports. 
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118773

laithsakka committed:
enable test_ellipsis_index_2 with Torch dynamo

[ghstack-poisoned]

lezcano reviewed (changes requested):
These tests work on eager, so they should work with dynamo, not with out eager implementation of NumPy. The fix should be implemented at a dynamo level.

All tests that are tagged with `xfailIfTorchDynamo` fail in exactly this category.

laithsakka commented:
@lezcano  this test fail when running with dynamo becuase of the check
        # `a[...]` was `a` in numpy <1.9.
        assert_(a[...].tensor._base is a.tensor)
     the output of dynamo is the output if np.array() which does not have the attribute tensor.
     so thats why I changed the test to force torch_.np.array() since those are not identical functions. 
     
     what do you suggest here, there is nothing to fix in dynmo?
     would it make sense to remove that check? or change it to (if the attribute tensor exist do  the check?)

lezcano commented:
Ah, I see. Perhaps we edited that test when bringing the test from NumPy. That line should be `a[...]._base is a._base`.

laithsakka commented:
I see I can change that 
this is the source of the reason. 
```
if TEST_WITH_TORCHDYNAMO:
    import numpy as np
    from numpy.testing import (
        assert_,
        assert_array_equal,
        assert_equal,
        assert_warns,
        HAS_REFCOUNT,
    )
else:
    import torch._numpy as np
    from torch._numpy.testing import (
        assert_,
        assert_array_equal,
        assert_equal,
        assert_warns,
        HAS_REFCOUNT,
    )
    ```

laithsakka commented:
> a[...]._base is a
that did not work. 
Traceback (most recent call last):
  File "/data/users/lsakka/pytorch/test/torch_np/numpy_tests/core/test_indexing.py", line 209, in test_ellipsis_index_2
    assert_(a[...]._base is a)
            ^^^^^^^^^^^^
AttributeError: 'ndarray' object has no attribute '_base'

lezcano commented:
Ah, in NumPy it's a public attribute. It should be `a[...].base is a`. I don't think we currently wire `base` to `_base` in dynamo, so that'd be the best fix here.

laithsakka commented:
> assert_(a[...].tensor._base is a.tensor)

how about
```
        # `a[...]` was `a` in numpy <1.9.
        if TEST_WITH_TORCHDYNAMO:
            assert_(a[...].base is a)
        else:
            assert_(a[...].tensor._base is a.tensor)
   ```
   
   I still do not think its a bug per say in dynamo since we are compiling two different functions in each path.

laithsakka committed:
Update on "Enable test_ellipsis_index_2 with Torch dynamo"

Fix issue #118819

test_ellipsis_index_2 is specifically testing properties of torch._numpy.array()
and that a field tensor is being added hence overriding the imports. 



[ghstack-poisoned]

lezcano commented:
We don't care all that much about being right on non dynamo. Note that we used these tests to test the implementation before the integration with dynamo, but now we don't care all that much about these tests being right on eager, as `torch._numpy` is not a public module.

```
assert_(a[...].base is a)
```
and xfailing when it's not dynamo SGTM

laithsakka commented:
> We don't care all that much about being right on non dynamo. Note that we used these tests to test the implementation before the integration with dynamo, but now we don't care all that much about these tests being right on eager, as `torch._numpy` is not a public module.
> 
> ```
> assert_(a[...].base is a)
> ```
> 
> and xfailing when it's not dynamo SGTM

I see I see so no need to run the test on eager and run it only in dynamo and expect failure on eager.

laithsakka committed:
Update on "Enable test_ellipsis_index_2 with Torch dynamo"

Fix issue #118819

test_ellipsis_index_2 is specifically testing properties of torch._numpy.array()
and that a field tensor is being added hence overriding the imports. 



[ghstack-poisoned]

laithsakka commented:
update test as agreed and discussed

lezcano reviewed (approved):
SGTM. Note that this feature is currently broken when dealing with graphbreaks, as we copy tensors when creating graph breaks. Fixing it would be a bit tricky I think, so it's good as is.

lezcano commented:
@pytorchbot merge
